aiida/repository/control.py

```diff
@@ -0,0 +1,81 @@
+"""Module for overall repository control commands."""
+# Note: these functions are not methods of `AbstractRepositoryBackend` because they need access to the orm.
+# This is because they have to go through all the nodes to gather the list of keys that AiiDA is keeping
+# track of (since they are descentralized in each node entry).
+# See the get_unreferenced_keyset function
+from aiida.manage.manager import get_manager
+from aiida.orm.implementation import Backend
+
+
+def repository_maintain(full: bool = False, backend: Backend = None, **kwargs) -> dict:
+    """Performs maintenance tasks on the repository."""
+    maintainance_report = {'database': {}, 'repository': {}, 'user_info': ''}
+
+    if backend is None:
+        backend = get_manager().get_backend()
+    repository_backend = backend.get_repository()
+
+    unreferenced_objects = get_unreferenced_keyset(aiida_backend=backend)
+
+    maintainance_report['repository']['unreferenced'] = len(unreferenced_objects)  # type: ignore
+    maintainance_report['repository']['info'] = repository_backend.get_info()  # type: ignore
+
+    repository_backend.delete_objects(list(unreferenced_objects))
+
+    # Perform the maintainance operations in the repository
+    maintainance_report['repository']['maintain'] = repository_backend.maintain(full=full, **kwargs)  # type: ignore
+
+    return maintainance_report
+
+
+def repository_info(statistics: bool = False, backend: Backend = None, **kwargs) -> dict:
+    """Returns relevant information on the repository."""
+    if backend is None:
+        backend = get_manager().get_backend()
+    repository_backend = backend.get_repository()
+    return repository_backend.get_info(statistics)
+
+
+def get_unreferenced_keyset(check_consistency: bool = True, aiida_backend: Backend = None, **kwargs) -> set:
+    """Returns the keyset of objects that exist in the repository but are not tracked by AiiDA.
+
+    This should be all the soft-deleted files.
+    """
+    from aiida import orm
+
+    if aiida_backend is None:
+        aiida_backend = get_manager().get_backend()
+
+    repository_backend = aiida_backend.get_repository()
+
+    keyset_backend = set(repository_backend.list_objects())
+    keyset_aiidadb = set(orm.Node.objects(aiida_backend).iter_repo_keys())
+
+    if check_consistency:
+        if len(keyset_aiidadb - keyset_backend) > 0:
+            raise RuntimeError('Database seems corrupted (some tracked objects are not in the repository). Aborting!')
+
+    return keyset_backend - keyset_aiidadb
+
+
+def get_repository_report(backend: Backend = None, **kwargs) -> dict:
+    """Performs a report on the status of the repository."""
+    report = {'user_info': ''}
+
+    if backend is None:
+        backend = get_manager().get_backend()
+    unreferenced_objects = get_unreferenced_keyset(aiida_backend=backend)
+
+    num_unref = len(unreferenced_objects)
+    report['user_info'] += f' > Unreferenced files to delete: {num_unref}'
```

Comment:
Think in these kinds of report methods that it is better to keep the formatting to a minimum. That should be the task of the caller who is going to print this to screen somehow.
```suggestion
    report['user_info'] += f'Unreferenced files to delete: {num_unref}'
```

Reply:
Well, there is an issue here again related to the inherent versatility of the repository. How can a generic caller like the CLI know how to format the info from the repository if this can be vastly different things? Having the underlying method return a pre-formated text that then the CLI can decide to print (or the ORM can ignore) was the simplest solution, as any other option would seem to require the design of some sort of "communication protocol" that is expected from the repository backend.

PS: Again, here you also need to take into account that you are looking at the line added by the `unreferenced_objects` count that we do on our end, but this `user_info` would also be passed to the repository backend to be populated with further info (like the space in MB these items occupy).

Reply:
I see the problem, but I still don't think that having these methods that spit out preformatted strings that are then just to be printed as a flexible solution. I think it is better to have a conditinal in the CLI to check the repo type and have it retrieve the information it wants to display and format it however it wants to, so that it is consistent. But if you really think this method is necessary, then I think the formatting should be as minimal as possible. This ` > ` formatting seems really unnecessary.

Reply:
> I think it is better to have a conditinal in the CLI to check the repo type and have it retrieve the information it wants to display and format it however it wants to, so that it is consistent.

But this basically breaks the modularity of the abstraction, no? The CLI needs to know about what information gets provided by all possible repository backends.

> But if you really think this method is necessary, then I think the formatting should be as minimal as possible. This `>` formatting seems really unnecessary.

Yeah, ok, that is a fair point, I can remove those.